libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/TelemetryWrapper.kt

```diff
@@ -0,0 +1,195 @@
+import com.mapbox.navigation.core.internal.telemetry.UserFeedbackCallback
+import com.mapbox.navigation.core.telemetry.events.FeedbackEvent
+import com.mapbox.navigation.core.telemetry.events.FeedbackMetadata
+import com.mapbox.navigation.core.telemetry.events.FeedbackMetadataWrapper
+import com.mapbox.navigation.metrics.MapboxMetricsReporter
+import com.mapbox.navigation.metrics.internal.TelemetryUtilsDelegate
+import com.mapbox.navigation.utils.internal.logD
+import java.util.concurrent.CopyOnWriteArrayList
+
+/**
+ * This wrapper is needed just for unit tests in order not to deal with types from Common SDK
+ */
+@UiThread
+internal class TelemetryStateWatcher(
+    private val telemetryUtils: TelemetryUtils = TelemetryUtils()
+) {
+
+    private val observers = CopyOnWriteArrayList<Observer>()
+
+    private val telemetryStateObserver = TelemetryCollectionStateObserver { state ->
+        observers.forEach {
+            it.onStateChanged(state.isEnabled)
+        }
+    }
+
+    fun registerObserver(observer: Observer) {
+        observers.add(observer)
+        if (observers.size == 1) {
+            telemetryUtils.registerTelemetryCollectionStateObserver(telemetryStateObserver)
+        }
+    }
+
+    fun unregisterObserver(observer: Observer) {
+        observers.remove(observer)
+        if (observers.isEmpty()) {
+            telemetryUtils.unregisterTelemetryCollectionStateObserver(telemetryStateObserver)
+        }
+    }
+
+    fun interface Observer {
+        fun onStateChanged(telemetryEnable: Boolean)
+    }
+
+    private companion object {
+        val TelemetryCollectionState.isEnabled: Boolean
+            get() = this == TelemetryCollectionState.ENABLED
+    }
+}
+
+/**
+ * Class that manages [MapboxNavigationTelemetry] initialisation. Listens to telemetry state events
+ * and can enable/disable telemetry in runtime.
+ *
+ * TODO(NAVAND-1820) ensure that we have 1-1 relationship between [TelemetryWrapper] and [MapboxNavigationTelemetry]
+ * [MapboxNavigationTelemetry] is very complex already and needs to be refactored.
+ */
+@UiThread
+internal class TelemetryWrapper(
+    private val telemetryStateWatcher: TelemetryStateWatcher = TelemetryStateWatcher()
+) {
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+    private lateinit var navigationOptions: NavigationOptions
+    private lateinit var userAgent: String
+
+    private var isWrapperInitialized = false
+    private var isTelemetryEnabled = false
+
+    private val telemetryStateObserver = TelemetryStateWatcher.Observer { isEnabled ->
+        if (!isTelemetryEnabled && isEnabled) {
+            initializeSdkTelemetry()
+        } else if (isTelemetryEnabled && !isEnabled) {
+            unInitializeSdkTelemetry()
+        }
+    }
+
+    fun initialize(
+        mapboxNavigation: MapboxNavigation,
+        navigationOptions: NavigationOptions,
+        userAgent: String
+    ) {
+        check(!isWrapperInitialized) {
```

Comment:
Do we really want to crash here?

Reply:
Yes, this is nice, but maybe let's not `if` it, but introduce an utility `assert` method that will only crash in debug. By the way, isn't there already something like that?

Reply:
> By the way, isn't there already something like that?

I haven't found anything like that. I've added check and failing only in debug builds